tools/system_libs.py

```diff
@@ -582,8 +600,11 @@ def maybe_noexcept(name):
     force_this = force_all or shortname in force
     if can_noexcept:
       shortname = maybe_noexcept(shortname)
-    if force_this:
-      suffix = 'bc' # .a files do not always link in all their parts; don't use them when forced
+    if force_this and not shared.Settings.WASM_OBJECT_FILES:
+      # .a files do not always link in all their parts; don't use them when forced
+      # When using wasm object files there are only .a archives but they get
```

Comment:
As an alternative to `--whole-archive`, we could link `.o` files into a larger combined `.o` file. That seems kind of simpler, and the `.o` would be a direct parallel to `.bc` for bitcode. What do you think?

Reply:
I did consider that approach.  However, I chose this approach for a couple of reasons.  Firstly, having libc.a be a collection of object files is more traditional, but easier (for me at least) to introspect and debug.  Secondly, incremental linking with `lld -r` is not a very common practice and while it is supported by the linker I am worried its less supported and could still have some issues.

In the long run I would like to find a way to avoid having to do this at all.  The reason its needed is because we don't know for sure that all the EXPORTED_FUNCTIONS come  from native code.  I would prefer to have a list of functions that are required at link time and include them via `--undefined=xxx`.    Then they will be automatically pulled out of the libraries as needed and also we can fail at lld time if they are missing.

Reply:
Makes sense, lgtm.